src/cryptography/hazmat/primitives/serialization/ssh.py

```diff
@@ -703,6 +718,262 @@ def _serialize_ssh_private_key(
     ed25519.Ed25519PublicKey,
 ]
 
+_SSH_CERT_PUBLIC_KEY_TYPES = typing.Union[
+    ec.EllipticCurvePublicKey,
+    rsa.RSAPublicKey,
+    ed25519.Ed25519PublicKey,
+]
+
+
+class SSHCertificateType(enum.Enum):
+    USER = 1
+    HOST = 2
+
+
+class SSHCertificate:
+    def __init__(
+        self,
+        _nonce: memoryview,
+        _public_key: _SSH_PUBLIC_KEY_TYPES,
+        _serial: int,
+        _cctype: int,
+        _key_id: memoryview,
+        _valid_principals: memoryview,
+        _valid_after: int,
+        _valid_before: int,
+        _critical_options: memoryview,
+        _extensions: memoryview,
+        _sig_type: memoryview,
+        _sig_key: memoryview,
+        _inner_sig_type: memoryview,
+        _signature: memoryview,
+        _tbs_cert_body: memoryview,
+        _cert_key_type: bytes,
+        _cert_body: memoryview,
+    ):
+        self._nonce = _nonce
+        self._public_key = _public_key
+        self._serial = _serial
+        try:
+            self._type = SSHCertificateType(_cctype)
+        except ValueError:
+            raise ValueError("Invalid certificate type")
+        self._key_id = _key_id
+        self._valid_principals = _valid_principals
+        self._valid_after = datetime.datetime.utcfromtimestamp(_valid_after)
+        self._valid_before = datetime.datetime.utcfromtimestamp(_valid_before)
+        self._critical_options = _critical_options
+        self._extensions = _extensions
+        self._sig_type = _sig_type
+        self._sig_key = _sig_key
+        self._inner_sig_type = _inner_sig_type
+        self._signature = _signature
+        self._cert_key_type = _cert_key_type
+        self._cert_body = _cert_body
+        self._tbs_cert_body = _tbs_cert_body
+
+    @property
+    def nonce(self) -> bytes:
+        return bytes(self._nonce)
+
+    def public_key(self) -> _SSH_CERT_PUBLIC_KEY_TYPES:
+        # make mypy happy until we remove DSA support entirely and
+        # the underlying union won't have a disallowed type
+        assert not isinstance(self._public_key, dsa.DSAPublicKey)
+        return self._public_key
+
+    @property
+    def serial(self) -> int:
+        return self._serial
+
+    @property
+    def type(self) -> SSHCertificateType:
+        return self._type
+
+    @property
+    def key_id(self) -> bytes:
+        return bytes(self._key_id)
+
+    @utils.cached_property
+    def valid_principals(self) -> typing.List[bytes]:
+        rest = self._valid_principals
+        principals = []
+        while rest:
+            principal, rest = _get_sshstr(rest)
+            principals.append(bytes(principal))
```

Comment:
Lazily parsing this means there could be syntax errors in it that we don't catch early, right? Is there a reason to be lazy?

Reply:
Yes there could be errors. The only reason to be lazy is to avoid the cost if not accessed. That said, the only expensive operation is loading a key so we could leave that one lazy and change the rest.